### PR TITLE
test: fix intermittent wallet_encryption failures on win64 task

### PR DESCRIPTION
**Pull request by jonatack**
- Fix intermittent failures like https://cirrus-ci.com/task/4571677035069440 and https://cirrus-ci.com/task/6566987122868224
- Document and tidy up RPC walletpassphrase MAX_SLEEP_TIME/nSleepTime
- Add test logging and an assertion and use named args

### CODE REVIEW

**Comment by mzumsande** [comment]
Strange that this happens twice in a day suddenly, without any recent changes that look related. The linked CI runs fail at different subtests, one at ("Check a timeout less than the limit") and one at ("Check a timeout greater than the limit") - so a fix should probably also affect both.

**Comment by MarcoFalke** [comment]
I can't see how this would happen unless the python clock is ahead of the system clock??

**Comment by jonatack** [comment]
Another one: https://cirrus-ci.com/task/4631236957175808

```
  File "C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\test\functional\wallet_encryption.py", line 93, in run_test
    assert_greater_than(expected_time_with_buffer, actual_time)
  File "C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\test\functional\test_framework\util.py", line 56, in assert_greater_than
    raise AssertionError("%s <= %s" % (str(thing1), str(thing2)))

AssertionError: 1751237054.993892 <= 1751237055
```

**Comment by samahanas** [comment]
How long did we fail, did we succeed, how long did the test fail, wrong,
wrong, how long did you get tired?

في الجمعة، 29 أبريل 2022 12:36 ص Jon Atack ***@***.***> كتب:

**Comment by samahanas** [comment]
I have a suggestion to remove all the memory of the old device and start
removing it again, perhaps this procedure will work

في الجمعة، 29 أبريل 2022 2:15 ص Martin Zumsande ***@***.***>
كتب:

**Comment by jonatack** [comment]
Another: https://cirrus-ci.com/task/6179044738727936

**Comment by vasild** [comment]
I think a possible explanation would be if the `GetTime()` in the C++ code on Windows ends up rounding _up_ to a whole second. So, for example:

1. expected time = int(10.5) + MAX = 10 + MAX
2. 0.1 second passes
3. walletpassphrase RPC is executed, asking it to keep the phrase for MAX seconds, it calculates now=ceil(10.6) + MAX = 11 + MAX // ceil() wtf!?
4. 0.1 second passes
5. expected with buffer = 10.7 + MAX // not truncated to whole seconds

So, we end up with actual time (11 + MAX) < expected with buffer (10.7 + MAX).

If this guess is correct, then the following should fix it:

```python
before = math.floor(time.time()) + MAX_VALUE
self.nodes[0].walletpassphrase(passphrase2, MAX_VALUE + 1000) 
after = math.ceil(time.time()) + MAX_VALUE 
actual = self.nodes[0].getwalletinfo()['unlocked_until'] 
assert_greater_than_or_equal(actual, before)
assert_greater_than_or_equal(after, actual)
```

**Comment by jonatack** [comment]
@vasild done, and added a commit for test logging (and named args and an assertion).

File: test/functional/wallet_encryption.py
Comment (vasild) [inline test/functional/wallet_encryption.py:77]: nit: maybe mention that this is the same as `MAX_SLEEP_TIME` in `src/wallet/rpc/encrypt.cpp`.

**Review by vasild — APPROVED** [review]
ACK 76f8dc9bd1b4efd543dc54c17a5f8e9d512580e1

**Comment by ryanofsky** [comment]
Just saw this error: https://cirrus-ci.com/task/4948737314258944?logs=functional_tests#L25 in #24230

Log:

```
 node0 2022-05-12T20:03:02.859261Z [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\src\rpc\request.cpp:179] [parse] ThreadRPCServer method=walletpassphrase user=__cookie__ 
 node0 2022-05-12T20:03:03.012258Z [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\src\rpc\server.cpp:530] [RPCRunLater] queue run of timer lockwallet(default_wallet) in 99999400 seconds (using HTTP) 
 test  2022-05-12T20:03:03.014000Z TestFramework (ERROR): Assertion failed 
                                   Traceback (most recent call last):
                                     File "C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\test\functional\test_framework\test_framework.py", line 133, in main
                                       self.run_test()
                                     File "C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\test\functional\wallet_encryption.py", line 85, in run_test
                                       assert_greater_than(expected_time_with_buffer, actual_time)
                                     File "C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\test\functional\test_framework\util.py", line 56, in assert_greater_than
                                       raise AssertionError("%s <= %s" % (str(thing1), str(thing2)))
                                   AssertionError: 1752385182.999136 <= 1752385183
```

**Comment by MarcoFalke** [comment]
Ok, then I think we should only maim the test on windows, not on all platforms.

**Review by vasild — APPROVED** [review]
ACK 1a82609ce078e2d3e5e242d85b275b950fd6fb8f

Why "maim"? The test functionality if fully preserved, no?

cc @achow101, author of these tests

**Comment by MarcoFalke** [comment]
> Why "maim"? The test functionality if fully preserved, no?

Sure, "maim" is a bit too strong. It is making the test lax by up to another second, while no one understands why.

**Comment by MarcoFalke** [comment]
Would be nice if someone could run this on msvc:

```cpp
#include <iostream>
#include <chrono>
#include <string>
 
template <typename TimePoint>
std::string to_string(const TimePoint& time_point)
{
    return std::to_string(time_point.time_since_epoch().count());
}
 
int main()
{
    using namespace std::literals::chrono_literals;
    using Sec = std::chrono::seconds;
 
    std::cout << "Time point\t" "Cast\t" "Floor\t" "Round\t" "Ceil\n";
    std::cout << "(ms)\t\t"     "(s)\t"  "(s)\t"   "(s)\t"   "(s)\n";
    for (const auto value_ms: {5432ms, 5678ms}) {
        std::chrono::time_point<std::chrono::system_clock, std::chrono::milliseconds>
            time_point_ms(value_ms);
 
        std::cout
            << to_string(time_point_ms) << "\t\t"
            << to_string(std::chrono::time_point_cast<Sec>(time_point_ms)) << "\t"
            << to_string(std::chrono::floor          <Sec>(time_point_ms)) << "\t"
            << to_string(std::chrono::round          <Sec>(time_point_ms)) << "\t"
            << to_string(std::chrono::ceil           <Sec>(time_point_ms)) << "\n";
    }
}
```

From https://en.cppreference.com/w/cpp/chrono/time_point/ceil. Should print:

```
Time point	Cast	Floor	Round	Ceil
(ms)		(s)	(s)	(s)	(s)
5432		5	5	5	6
5678		5	5	6	6

**Comment by MarcoFalke** [comment]
Unrelated, it might be good to not use mockable time here. libevent isn't affected by `setmocktime`, so using mockable time here is harmless but still a bit confusing:

```diff
diff --git a/src/util/time.h b/src/util/time.h
index 72956ea0d7..8d160c312e 100644
--- a/src/util/time.h
+++ b/src/util/time.h
@@ -18,6 +18,8 @@ using SteadySeconds = std::chrono::time_point<std::chrono::steady_clock, std::ch
 using SteadyMilliseconds = std::chrono::time_point<std::chrono::steady_clock, std::chrono::milliseconds>;
 using SteadyMicroseconds = std::chrono::time_point<std::chrono::steady_clock, std::chrono::microseconds>;
 
+using SysSeconds = std::chrono::time_point<std::chrono::system_clock, std::chrono::seconds>; // std::chrono::sys_seconds (C++20)
+
 void UninterruptibleSleep(const std::chrono::microseconds& n);
 
 /**
diff --git a/src/wallet/rpc/encrypt.cpp b/src/wallet/rpc/encrypt.cpp
index 802cc63d6d..8a663e71bd 100644
--- a/src/wallet/rpc/encrypt.cpp
+++ b/src/wallet/rpc/encrypt.cpp
@@ -75,7 +75,7 @@ RPCHelpMan walletpassphrase()
 
         pwallet->TopUpKeyPool();
 
-        pwallet->nRelockTime = GetTime() + nSleepTime;
+        pwallet->nRelockTime = Now<SysSeconds>().time_since_epoch().count() + nSleepTime;
         relock_time = pwallet->nRelockTime;
     }

**Comment by vasild** [comment]
> It is making the test lax by up to another second

That is true, but the point of the test is to ensure that the time limit is honored - 1. cannot be negative, 2. is capped at some limit and 3. if within the limits, then is accepted. The original test in 134cdc7cee3da7c554e40ad947a9cdcbb3069f13 contained 5 seconds buffer, so the point of the test is not to check that the value is accepted up to the second, exactly.

File: test/functional/wallet_encryption.py
Comment (vasild) [inline test/functional/wallet_encryption.py:86]: Would this print the values if the assertion fails?

```
>>> a=1
>>> b=2
>>> c=3
>>> assert a <= b <= c
>>> 
>>> c=0
>>> assert a <= b <= c
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
>>> 
```

If not, then maybe revert back to using `assert_greater_than()` because seeing the values in CI logs is so precious in debugging this.

**Comment by MarcoFalke** [comment]
> so the point of the test is not to check that the value is accepted up to the second, exactly.

Right. I don't care about this test. `nRelockTime` is only there for aesthetic reasons, it can't be relied upon anyway, since it doesn't take into account how long it takes to register the event in libevent. (If that takes 5 seconds, the value is off by 5 seconds)

I mostly worry the clock on windows doesn't work as expected and it will lead to more severe issues down the line (extremely unlikely, but worst case a consensus split)

**Comment by sipsorcery** [comment]
Visual Studio 2022 v17.1.6 9 (C++ compiler version 19.31.31107 for x86)

````
Time point      Cast    Floor   Round   Ceil                                                                          
(ms)            (s)     (s)     (s)     (s)                                                                             
5432            5       5       5       6                                                                             
5678            5       5       6       6
````

File: test/functional/wallet_encryption.py
Comment (MarcoFalke) [inline test/functional/wallet_encryption.py:84]: This doesn't seem right. If the Bitcoin Core clock was doing `ceil`, we'd see a failure rate of about 100%. `round` seems unlikely as well, as the failure rate would be about 50%.

Statistically, the error on windows should be 10-100ms, which can't be explained by cast/round/ceil.

Comment (vasild) [reply test/functional/wallet_encryption.py:84]: Well, this just ignores sub-second fractions and accounts for the time passed between the calls, keeping the original intention of the test (I hope!). Do you think it could still fail on Windows?

**Comment by MarcoFalke** [comment]
An alternative to this would be to use mocktime to set the exact seconds and then check the exact result. (Obviously mocktime couldn't shorten or extend the actual locking duration, but the test doesn't check that anyway).

**Comment by jonatack** [comment]
Will update to address feedback.

**Comment by jonatack** [comment]
Again https://cirrus-ci.com/task/4705957107204096

**Comment by jonatack** [comment]
https://cirrus-ci.com/task/4707637714485248

**Comment by MarcoFalke** [comment]
Is this still happening after the windows upgrade?

**Comment by MarcoFalke** [comment]
I haven't seen it yet, so it indeed seems like a bug in Windows 2019

**Comment by jonatack** [comment]
On master https://cirrus-ci.com/task/4974085586288640?logs=functional_tests#L45

**Comment by vasild** [comment]
https://cirrus-ci.com/task/4856438647422976?logs=functional_tests#L45

**Comment by jonatack** [comment]
https://cirrus-ci.com/task/5509434167263232?logs=functional_tests#L43

**Comment by jonatack** [comment]
Seems this won't be merged.

**Comment by MarcoFalke** [comment]
The reason I haven't merged this yet is that it has only one ACK, but outstanding review comments such as https://github.com/bitcoin/bitcoin/pull/25021#discussion_r872401870 and https://github.com/bitcoin/bitcoin/pull/25021#discussion_r873198183

Also, I am not sure if a test fix should modify the cpp source code. No objection to those changes, but maybe try to submit them separate from a test fix?

**Comment by vasild** [comment]
The reason for closing - "this won't be merged" seems unjustified to me. If Marco did not merge it in the past, that does not mean he will not merge it in the future. Or that another maintainer will not merge it in the future.

Indeed it has just one ACK. @achow101, maybe you can review this, since you are the original author of the tests.

I think this PR should be reopened, IMO it is an improvement even if it does not fix the mysterious test failures.

**Comment by MarcoFalke** [comment]
For reference, I created https://github.com/bitcoin/bitcoin/pull/25750 two weeks ago

**Comment by jonatack** [comment]
> I think this PR should be reopened, IMO it is an improvement even if it does not fix the mysterious test failures.

Ok, re-opening.

**Comment by jonatack** [comment]
> For reference, I created #25750 two weeks ago

As seen above, I'm happy to update this PR with suggestions for improvements or a different approach.

File: test/functional/wallet_encryption.py
Comment (jonatack) [reply test/functional/wallet_encryption.py:86]: Looking.

Comment (jonatack) [reply test/functional/wallet_encryption.py]: Good point. Updated the assertion error to print i.e. `AssertionError: walletpassphrase "unlocked_until" value of 1760141911 not in expected range of 1760141909..1760141910`

**Comment by jonatack** [comment]
Updated to improve assertion error, changed the first commit message from "fix" to "work around" and rebased to current master to enable clang-tidy iwyu checks on the second commit (will re-push after the clang-tidy CI run). Happy to split out the some of changes to a separate pull or to update the approach if reviewers prefer.

**Comment by fanquake** [comment]
> https://github.com/bitcoin/bitcoin/pull/25021/files#r873198183

Has this been is addressed? If yes, can you mark it as resolved, if not, can you comment?

> changed the first commit message from "fix" to "work around" 

If this is now a work around, rather than a fix, I'd rather use the work-around that is contained to the platform that is broken, i.e #25750, rather than refactoring code that works for all other platforms, and essentially hiding the bug on Windows. At least until we properly understand the issue; see debugging in https://github.com/bitcoin/bitcoin/issues/25482.

The PR title and description need updating in any case.

**Comment by MarcoFalke** [comment]
Can you explain why the workaround here is better than the one in https://github.com/bitcoin/bitcoin/pull/25750?

**Comment by jonatack** [comment]
> Can you explain why the workaround here is better than the one in #25750?

I'm happy to update this PR with suggestions for improvements or a different approach.

ATM am looking into the discussion in https://github.com/bitcoin/bitcoin/issues/25482.

**Comment by achow101** [comment]
It seems like instead of messing around with buffers and relaxing ranges, this could just be resolved through the use of mock time so that bitcoind is set to a fixed time.

**Comment by jonatack** [comment]
I'll give that a shot 👍

**Comment by MarcoFalke** [comment]
For reference, I suggested mocktime in May: https://github.com/bitcoin/bitcoin/pull/25021#issuecomment-1132673747

This should be fine to do here, as the time returned has no guarantee to be related to the real relock time, which happens completely independent in the libevent scheduler thread.

**Pull request closed without merging**
